app-backend/tile/src/main/scala/tool/TileSources.scala

```diff
@@ -101,26 +102,78 @@ object TileSources extends LazyLogging {
 
     case ProjectRaster(projId, Some(band), maybeND) => {
       Mosaic.rawForExtent(projId, zoom, Some(Projected(extent.toPolygon, 3857)))
-        .map({ tile => tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)) }).value
+        .map({ tile =>
+          TileProvider(tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)), None)
+        }).value
     }
 
     case _ => Future.successful(None)
   }
 
   /** This source provides support for z/x/y TMS tiles */
-  def cachedTmsSource(r: RFMLRaster, z: Int, x: Int, y: Int)(implicit database: Database): Future[Option[Tile]] =
+  def cachedTmsSource(r: RFMLRaster, buffer: Boolean, z: Int, x: Int, y: Int)(implicit database: Database): Future[Option[TileProvider]] =
     r match {
       case scene @ SceneRaster(sceneId, Some(band), maybeND) =>
-        LayerCache.layerTile(sceneId, z, SpatialKey(x, y))
-          .map({ tile => tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)) }).value
+        if (buffer) {
+          val provider: OptionT[Future, TileProvider] = for {
+            tl <- LayerCache.layerTile(sceneId, z, SpatialKey(x - 1, y - 1))
+                    .map({ tile => tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)) })
+            tm <- LayerCache.layerTile(sceneId, z, SpatialKey(x, y - 1))
+                    .map({ tile => tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)) })
+            tr <- LayerCache.layerTile(sceneId, z, SpatialKey(x + 1 , y - 1))
+                    .map({ tile => tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)) })
+            ml <- LayerCache.layerTile(sceneId, z, SpatialKey(x - 1, y))
+                    .map({ tile => tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)) })
+            mm <- LayerCache.layerTile(sceneId, z, SpatialKey(x, y))
+                    .map({ tile => tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)) })
+            mr <- LayerCache.layerTile(sceneId, z, SpatialKey(x + 1, y))
+                    .map({ tile => tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)) })
+            bl <- LayerCache.layerTile(sceneId, z, SpatialKey(x - 1, y + 1))
+                    .map({ tile => tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)) })
+            bm <- LayerCache.layerTile(sceneId, z, SpatialKey(x, y + 1))
+                    .map({ tile => tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)) })
+            br <- LayerCache.layerTile(sceneId, z, SpatialKey(x + 1, y + 1))
+                    .map({ tile => tile.band(band).interpretAs(maybeND.getOrElse(tile.cellType)) })
+          } yield TileProvider(mm, Some(Buffers(tl, tm, tr, ml, mr,bl, bm, br)))
```

Comment:
Since each call to `LayerCache.layerTiles` is an `OptionT` does it mean we're dropping those tiles which do not have neighbors on all sides?